plugin/healthchecker/filter.go

```diff
@@ -0,0 +1,116 @@
+package healthchecker
+
+import (
+	"time"
+
+	"github.com/bluele/gcache"
+	"github.com/miekg/dns"
+)
+
+type (
+	HealthCheckFilter struct {
+		cache    gcache.Cache
+		checker  Checker
+		interval time.Duration
+		names    map[string]struct{}
+	}
+
+	entry struct {
+		dnsRecord dns.RR
+		healthy   bool
+		quit      chan struct{}
+	}
+)
+
+func NewHealthCheckFilter(checker Checker, size, interval int, names map[string]struct{}) *HealthCheckFilter {
+	return &HealthCheckFilter{
+		cache: gcache.New(size).LRU().EvictedFunc(func(key interface{}, val interface{}) {
+			if e, ok := val.(entry); ok {
+				close(e.quit)
+			}
+		}).Build(),
+		checker:  checker,
+		interval: time.Duration(interval) * time.Millisecond,
+		names:    names,
+	}
+}
+
+func (p *HealthCheckFilter) FilterRecords(records []dns.RR) []dns.RR {
+	if p.cache.Len(false) == 0 {
+		recs := make([]dns.RR, len(records))
+		copy(recs, records)
+		go func() {
+			for _, r := range recs {
+				if _, ok := p.names[r.Header().Name]; ok || len(p.names) == 0 {
+					p.put(r)
+				}
+			}
+		}()
+		return records
+	}
+
+	result := make([]dns.RR, 0, len(records))
+
+	for _, r := range records {
+		if _, ok := p.names[r.Header().Name]; ok || len(p.names) == 0 {
+			e := p.get(r.String())
+			if e != nil {
+				if e.healthy {
+					result = append(result, e.dnsRecord)
+				}
+				continue
+			}
+			p.put(r)
+		}
+		result = append(result, r)
+	}
+
+	return result
+}
+
+func (p *HealthCheckFilter) put(rec dns.RR) {
+	health := p.checker.Check(rec)
+	quit := make(chan struct{})
+	err := p.cache.Set(rec.String(), entry{
+		dnsRecord: rec,
+		healthy:   health,
+		quit:      quit,
+	})
+	if err != nil {
+		log.Warningf("could not put an entry to cache: %s", rec.String())
+		return
+	}
+
+	ticker := time.NewTimer(p.interval)
+	go func() {
+		for {
+			select {
+			case <-quit:
+				return
+			case <-ticker.C:
+				err := p.cache.Set(rec.String(), entry{
+					dnsRecord: rec,
+					healthy:   p.checker.Check(rec),
+					quit:      quit,
+				})
```

Comment:
What will happen if  we try to set updated record when `quit` is already closed? It seems we will have record in the cache with closed channel and the next time eviction cause `panic` because of trying to close already closed channel. I suppose we should check record existing every tick at first.

Reply:
Could reproduce on this code:
```
type tmpcheck struct {
}

func (t *tmpcheck) Check(record dns.RR) bool {
	return true
}

func TestPanic(t *testing.T) {
	checker := &tmpcheck{}

	f := NewHealthCheckFilter(checker, 2, 200, map[string]struct{}{})

	a := &dns.A{A: net.ParseIP("127.0.0.1")}
	a2 := &dns.A{A: net.ParseIP("127.0.0.2")}
	a3 := &dns.A{A: net.ParseIP("127.0.0.3")}
	a4 := &dns.A{A: net.ParseIP("127.0.0.4")}

	f.put(a)
	f.put(a2)
	time.Sleep(200 * time.Millisecond)
	f.put(a3)
	time.Sleep(200 * time.Millisecond)
	f.put(a4)

	time.Sleep(time.Second)
}
```

Reply:
> I suppose we should check record existing every tick at first.

Even if you check key existence before `p.cache.Set`, eviction can happen right after it. 
1. LookUp (routine 1)
2. Eviction (routine 2)
3. Set (routine 1)

Mutex will prevent eviction during update. 

Also `gcache` indeed doesn't have lookup function. Maybe [golang-lru](https://github.com/hashicorp/golang-lru/blob/80c98217689d6df152309d574ccc682b21dc802c/lru.go#L98) will be a better choice, we use that in NeoFS.

Reply:
Yes, we decided checking if channel is closed. But it seems this doesn't solve the problem 
/cc @masterSplinter01